[RyanNgWH] iP

RyanNgWH opened the pull request:
# :notebook_with_decorative_cover: Duke - High performance tasks manager solution

> [!CAUTION]
> - This project is under **active** development
> - There will be bugs and breaking changes
> - **Do not use this app as the sole way of storing important tasks**

Duke is a simple tasks command line based tasks manager developed with speed as the main focus.

## Features
* Text-based
* Low learning curve
* FAST to use

## Getting Started
1. Grab the latest release from [here](https://github.com/RyanNgWH/ip/releases)
1. Launch the application
1. Add your tasks
1. Profit! 
---
Alternatively, you could launch it from the command line with the following command:
```bash
java -jar duke.jar
```

> [!WARNING]
> This project was developed with `Java 11` and is not guaranteed to work with other versions of Java

## Supported Commands
```bash
usage:    <command> <arguments>
commands:
    bye                                   # Exit the application
    list      [/date]                     # List all tasks in the application (with an optional date filter - in the format YYYY/MM/DD
    mark      {index}                      # Mark the task as completed
    unmark    {index}                     # Mark the task as not completed
    delete    {index}                     # Delete the task
    todo      {description}               # Create a todo with the specified description
    deadline  {description} {/by}         # Create a deadline with the specified description and deadline
    event     {description} {/from} {/to} # Create an event with the specified description and duration
```

## Roadmap
- [X] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

RyanNgWH committed:
Implement initial skeletal chatbot

- Rename chatbot
- Add greeting
- Add exit message

RyanNgWH committed:
Change project structure

- Create Duke package
- Create UI package
- Move CLI UI code to UI package

RyanNgWH committed:
Improve skeletal chatbot

- Echo commands entered by user
- Exits on `bye`

RyanNgWH committed:
Create custom storage full exception

RyanNgWH committed:
Create storage class

- Add method to store items
- Add method to list stored items

RyanNgWH committed:
Fix list numbering issue

Fix issue where list was numbering every entry as '1'

RyanNgWH committed:
Implement UI for adding and listing items

- Add UI logic for storing text from user input
- Add UI logic for listing stored user text

RyanNgWH committed:
Add mark functionality

- Add method to mark item in storage
- Add method to unmark item in storage

RyanNgWH committed:
Refactor storage of tasks

- Add new Task class
- Migrate storage to store Tasks instead of primitives

RyanNgWH committed:
Update UI to allow marking and unmarking of tasks

RyanNgWH committed:
Add todo task type

RyanNgWH committed:
Fix typo in todo class comment

RyanNgWH committed:
Create deadline task type

RyanNgWH committed:
Fix typos in the deadline class

- Add missing parameter field in constructor javadoc

RyanNgWH committed:
Create event task type

RyanNgWH committed:
Create exception for missing arguments in cli

RyanNgWH committed:
Create task not supported exception

- Fix typo in storage full exception class

RyanNgWH committed:
Implement string representation of specific tasks

- Add toString method for Deadline, Event & Todo
- Fix typos

RyanNgWH committed:
Implement storage of different tasks

- Update UI to support storing of different tasks

RyanNgWH committed:
Change UI message when adding task to list

RyanNgWH committed:
Modify runtest.sh script to match project structure

RyanNgWH committed:
Create test case input and expected output files

RyanNgWH committed:
Create application specific exception

- Migrate all current exceptions to DukeExceptions

RyanNgWH committed:
Add exception for empty todo description

RyanNgWH committed:
Migrate storage implementation to ArrayList

- Remove unnecesary exception

RyanNgWH committed:
Fix runtest.sh

RyanNgWH committed:
Implement deletion of task

- Add method to delete item from storage
- Remove unnecessary exception catches
- Add exception handling for mark, unmark & delete methods

RyanNgWH committed:
Update automated i/o testing files

- Add delete method to input and expected output files

RyanNgWH committed:
Change tab size from 2 to 4

RyanNgWH committed:
Update switch case indentation

Indentation for case should align with switch statement

Create project formatter override for eclipse formatting

RyanNgWH committed:
Add vscode configuration to gitignore

RyanNgWH committed:
Add third-party JSON package

RyanNgWH committed:
Add getters for each task field & Add enums for task type

JSONObject needs public getters to automatically obtain Object fields

Enums stadardises the task type to minimise errors in the future

RyanNgWH committed:
Add method to save tasks to JSON file

RyanNgWH committed:
Add data files to gitignore

RyanNgWH committed:
Add constructors that include task status for all tasks

Tasks should be able to be created without default status value of false

RyanNgWH committed:
Implement load stored tasks from file

Application should be able to load stored tasks from file

RyanNgWH committed:
Add exception handling for corrupted and unredable tasks data file

RyanNgWH committed:
Change corrupted file warning message

RyanNgWH committed:
Merge branch 'branch-Level-7'

RyanNgWH committed:
Add new invalid argument exception

Modify exception javadocs to be more descriptive

RyanNgWH committed:
Refactor deadline task to store due date as Instant instead of string

Due date input should be sanitised to only accept legitimate dates instead of any string

RyanNgWH committed:
Add error handling of invalid deadline due date format & storing of date in file

RyanNgWH committed:
Add exception for corrupted file

Empty file and malformed json fields should be handled gracefully

RyanNgWH committed:
Refactor event start and end date storage type

Event should recognise date and time instead of accepting any strings

RyanNgWH committed:
Modify UI datetime to be in a more readable format

RyanNgWH committed:
Add functionality to list events/deadlines on a specified date

RyanNgWH committed:
Merge branch 'branch-Level-8'

RyanNgWH committed:
Refactor UI classes

Other methods of UI other than the CLI should be possible

Let's create and have different UIs implement an abstract UI class

RyanNgWH committed:
Refactor Main class to use new UI structure

RyanNgWH committed:
Create parser class for making sense of user commands

RyanNgWH committed:
Create missing task and corrupted task exceptions

Exceptions should specifically mention what is the issue

RyanNgWH committed:
Add exception handling to user date to instant parser

RyanNgWH committed:
Refactor task handling from storage to task list class

Tasks should be handled independently of storage

RyanNgWH committed:
Refactor application to use new project structure

Application should use new structure which decoupled storage and tasklist

RyanNgWH committed:
Refactor Duke main class

RyanNgWH committed:
Create abstract command class

Commands can be abstracted

RyanNgWH committed:
Create exit command

RyanNgWH committed:
Add exception for invalid command

RyanNgWH committed:
Add duke exception handling in command execute method & Fix typos

RyanNgWH committed:
Create list command

RyanNgWH committed:
Refactor CLI UI & parser to use new command classes

RyanNgWH committed:
Fix list display showing additional empty line

RyanNgWH committed:
Add storage exception & fix typos

RyanNgWH committed:
Add mark command & Refactor CLI/Parser to use new command

RyanNgWH committed:
Change IOException to Storage exception for file saving method

Error saving to file is a duke applicaiton exception and should be handled as one

RyanNgWH committed:
Refactor unmark logic

Unmarking of tasks can be accomplished using the mark command

RyanNgWH committed:
Refactor delete command to use command class

RyanNgWH committed:
Refactor add task functionality to use command class & Refactor save to file function in taskList

RyanNgWH committed:
Merge branch 'branch-A-MoreOOP'

RyanNgWH committed:
Merge branch 'add-gradle-support'

RyanNgWH committed:
Add local settings to git ignore

RyanNgWH committed:
Configure gradle

Add checkstyle plugin to gradle & JSON package dependency

RyanNgWH committed:
Fix styles using checkstyle

RyanNgWH committed:
Add test cases for delete command class & Fix formatting issues

RyanNgWH committed:
Add test cases for add command class

RyanNgWH committed:
Update gitignore & local settings

RyanNgWH committed:
Add test cases for the exit command class

RyanNgWH committed:
Rename test case in exit command test

RyanNgWH committed:
Fix list empty tasks list bug crashes program

RyanNgWH committed:
Fix application crash on invalid date bug in list command

RyanNgWH committed:
Fix typos in comments

RyanNgWH committed:
Add test cases for list command class

RyanNgWH committed:
Fix formatting error in mark command output

RyanNgWH committed:
Add test cases for mark command class

RyanNgWH committed:
Add equals method to each custom class

Each custom class object should have a proper way to compare to each other

RyanNgWH committed:
Fix null date bug in list command class equals method

RyanNgWH committed:
Refactor input parsing

Parser should be able to handle additional unnecessary whitespaces between commands/arguments

RyanNgWH committed:
Fix uncaught exceptions in parser & formatting issues

RyanNgWH committed:
Add test cases for parser class

RyanNgWH committed:
Add test cases for deadline class

RyanNgWH committed:
Add test cases for event class

RyanNgWH committed:
Merge branch 'branch-A-JUnit'

RyanNgWH committed:
Fix checkstyles

RyanNgWH committed:
Fix code styles and javadoc

RyanNgWH committed:
Fix test cases checkstyles

RyanNgWH committed:
Add tasklist method to filter tasks by keyword

RyanNgWH committed:
Add missing javadoc parameter

RyanNgWH committed:
Add find command class

RyanNgWH committed:
Add parser case for "find" command

RyanNgWH committed:
Fix keyword not being found bug

Keyword should filter output instead of other way round

RyanNgWH committed:
Add test cases for parser "find" command

RyanNgWH committed:
Add test cases for find command class

RyanNgWH committed:
Fix typos in javadoc

RyanNgWH committed:
Merge branch 'branch-A-JavaDoc'

RyanNgWH committed:
Merge branch 'branch-A-CodingStandard'

RyanNgWH committed:
Merge branch 'branch-Level-9'

RyanNgWH committed:
Fix checkstyles

RyanNgWH committed:
Add java version to gradle

RyanNgWH committed:
Add javafx to gradle

RyanNgWH committed:
Change start UI method to allow initialisation arguments

RyanNgWH committed:
Create initial GUI layout with styling

RyanNgWH committed:
Add functionality to execute commands without printing to command line

RyanNgWH committed:
Add initial GUI functionality to respond to commands

RyanNgWH committed:
Add styling to GUI using dialog boxes

smolegz commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,71 @@
+     * @param taskList Tasklist used for the command
+     * @param ui       UI used for the command
+     */
+    @Override
+    public void execute(TaskList taskList, Ui ui) throws DukeException {
+        // Add task
+        taskList.addTask(task);
+
+        // Save to file
+        taskList.saveTasks();
+
+        // Print success message
+        System.out.println("Got it. I've added this task:");
+        System.out.println(String.format("  %s", task.toString()));
+        System.out.println(String.format("Now you have %d tasks in the list.", taskList.size()));
+    }
+
+    /**
+     * Indicates whether some other object is "equal to" this command
+     *
+     * @param obj Object to be checked against
+     *
+     * @return True if equal, False otherwise
+     */
```
You can use the @inheritDoc to help you generate your JavaDoc easier. [Example of using @inheritDoc](https://www.baeldung.com/javadoc-see-vs-link)

RyanNgWH replied:
Will do! Thanks for letting me know!

smolegz commented on src/main/java/duke/storage/Deadline.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @return True if deadline is on the specified date, false otherwise
+     */
+    public boolean isOn(Instant date) {
```
Would this method be clearer if it was renamed to isSameDate()?

RyanNgWH replied:
That would indeed be a better naming convention. Thank you for the suggestion!

smolegz commented on src/main/java/duke/exceptions/InvalidCommandException.java:
```diff
@@ -0,0 +1,18 @@
+    /**
+     * Creates a InvalidCommandException
+     *
+     * @param errorMessage Error message
```
I think you can omit out the @param for such method since the parameter name is self-explanatory 😄.

smolegz commented on src/main/java/duke/storage/Deadline.java:
```diff
@@ -0,0 +1,111 @@
+     * @param dueDate     Due date of the deadline
+     * @param isDone      Status of the deadline
+     */
+    public Deadline(String description, Instant dueDate, boolean isDone) {
```
Noticed that isDone and isCompleted means refers to the status of a task. Would it be better to uniformly name these two Boolean variables?

RyanNgWH replied:
Good catch! Will make the relevant changes

smolegz reviewed (commented):
Looks very good to me! Classes and methods were well-documented and commented, which made the reading experience seamless and understandable. Definitely learnt a few things from you!

RyanNgWH committed:
Fix formatting error

RyanNgWH committed:
Refactor GUI rendeing to use FXML file

RyanNgWH committed:
Add formatting to dialogbox

RyanNgWH committed:
Merge branch 'branch-Level-10'

owenong1 commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,85 @@
+ *
+ * @author Ryan NgWH
+ */
+public class AddCommand extends Command {
```
Perhaps a more descriptive name would be AddTaskCommand, since in the future there may be other things you could be adding.

owenong1 commented on src/main/java/duke/commands/Command.java:
```diff
@@ -0,0 +1,74 @@
+     *
+     * @return String containing the output of the command
+     */
+    public abstract String execute(TaskList taskList) throws DukeException;
```
I feel like overloading 'execute' may not be very representative of the functionality. Perhaps a name like getString or getOutputString would be more accurate

RyanNgWH replied:
Good point. Will consider your alternatives!

owenong1 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,232 @@
+     */
+    public static Instant userDateToInstant(String date, String time)
+            throws NumberFormatException, StringIndexOutOfBoundsException, DateTimeException {
+        return LocalDateTime.of(
```
I think you can explore the use of the .split function to split by the '/' character. it may be less wordy + more evolvable than substring.

owenong1 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,232 @@
+            }
+
+        case "mark": // Mark task
+            if (splitInput.size() <= 1) {
```
There should be some better logic for handling size <= 1 instead of repeating it in majority of the cases. Maybe another nested loop? It would improve readability.

RyanNgWH replied:
Will look into it. Thank you for the suggestion!

owenong1 reviewed (commented):
LGTM. Mostly just some naming/structural issues but otherwise the logic isfine. Good job!!

RyanNgWH committed:
Add assertions

Assertions are missing throughout code

Assertions help to define assumptions about the program state and can be used to identify possible bugs

Let's,
* Add assertion in GUI class to ensure a valid command is returned
* Add assertion in storage class to ensure file to be saved to is writable
* Add assertions in storage class to ensure no null task is written to the program
* Add assertions in Duke class to ensure all required objects are instantiated before running the program

RyanNgWH committed:
Enable assertions in gradle

RyanNgWH committed:
Refactor event ecompasses method

Return statement is complicated

Return value has 3 boolean expressions and can be difficult to evaluated in one statement

Let's split the statement into 3 intermediate values before calculating the final statement

RyanNgWH committed:
Change parameter name of deadline constructor

Parameter uses name that is not standardised with other classes

Standardising parameter naming makes it easier for other developers to understand what the parameter is used for

RyanNgWH committed:
Add whitespace

RyanNgWH committed:
Refactor getTasks to use streams

RyanNgWH committed:
Merge pull request #1 from RyanNgWH/branch-A-Assertions

Add Assertions

RyanNgWH committed:
Merge branch 'master' into branch-A-CodeQuality

RyanNgWH committed:
Merge pull request #2 from RyanNgWH/branch-A-CodeQuality

Improve Code Quality

RyanNgWH committed:
Merge branch 'master' into branch-A-Streams

RyanNgWH committed:
Merge pull request #3 from RyanNgWH/branch-A-Streams

Use streams

RyanNgWH committed:
Fix checkstyles

RyanNgWH committed:
Add Continuous Integration (CI)

RyanNgWH committed:
Fix test cases overwriting application save file

RyanNgWH committed:
Set timezone of application to UTC+8

RyanNgWH committed:
Fix timezone to UTC+8

RyanNgWH committed:
Fix missing directory error in test cases environment setup

RyanNgWH committed:
Fix timezone of date to instant parser to UTC+8

RyanNgWH committed:
Fix checkstyles

RyanNgWH committed:
Rename isDone to isCompleted to standardise naming conventions

RyanNgWH committed:
Change tasks to support visibility (archived)

RyanNgWH committed:
Remove redundant save to file method calls

RyanNgWH committed:
Add archive functionality to taskList

RyanNgWH committed:
Add methods to archive and unarchive tasks

RyanNgWH committed:
Add archive command

RyanNgWH committed:
Update commands to support archived tasks

RyanNgWH committed:
Update tests to support archived tasks

RyanNgWH committed:
Update parser to support archived tasks

RyanNgWH committed:
Update tasklist to support new add task saving method

RyanNgWH committed:
Update load from file method to supoprt archived tasks

RyanNgWH committed:
Fix bug in getTasksString method

Stream is using wrong list to generate the index which is causing an index out of range exception

RyanNgWH committed:
Reset test data file in test cases

File is not reset in some test cases which causes bugs in other test cases

RyanNgWH committed:
Add parser method for archive command

RyanNgWH committed:
Add list archived tasks functionality to parser

RyanNgWH committed:
Fix formatting

RyanNgWH committed:
Add functionality to save archived tasks to file

RyanNgWH committed:
Add unarchive method to parser

RyanNgWH committed:
Revert UI back to GUI

RyanNgWH committed:
Fix checkstyles

RyanNgWH committed:
Merge branch 'branch-BCD-Extension'

RyanNgWH committed:
Update gradle openJFX configuration

RyanNgWH committed:
Add customisations to dialog box

RyanNgWH committed:
Add background color to main window

RyanNgWH committed:
Rename application to `Ciara`

RyanNgWH committed:
Add login welcome message

RyanNgWH committed:
Add functionality to stop execution on bye command with GUI

RyanNgWH committed:
Merge branch 'branch-A-Personality'

RyanNgWH committed:
Add test cases for archive command execution

RyanNgWH committed:
Add parser test cases for listing archived tasks

RyanNgWH committed:
Merge branch 'branch-A-MoreTesting'

RyanNgWH committed:
Update README

RyanNgWH committed:
Update README & Github pages

RyanNgWH committed:
Update Github pages theme

RyanNgWH committed:
Update github pages title and description

RyanNgWH committed:
Remove unnecessary comments

RyanNgWH committed:
Split parser method into smaller methods

RyanNgWH committed:
Fix header comments

RyanNgWH committed:
Fix checkstyles

RyanNgWH committed:
Exclude windows from some test cases

Test cases fail on windows OS while passing on others

Temporarily exclude test cases while issue is being investigate

RyanNgWH committed:
Exclude errernous test case from windows

RyanNgWH committed:
Set title for GUI

RyanNgWH committed:
Add javafx version to gradle
